File: scratch/inspect_all_templates.py
import os
import zipfile
import xml.etree.ElementTree as ET

templates = [
    "1. Mau 10_KND_Ban tu kiem diem_DHKT_2025.doc",
    "1. Nghị quyết LCĐ.docx",
    "1. Nghị quyết Đoàn Trường (02 bản).docx",
    "3. Biên bản họp Liên chi Đoàn.docx",
    "4. Nghị quyết Chi Đoàn.docx",
    "5. Biên bản họp Chi Đoàn.docx",
    "6. Biên bản họp lớp.docx"
]

public_dir = r"D:\CBSV\public"
output_file = r"D:\CBSV\scratch\templates_dump.txt"

namespaces = {
    'w': 'http://schemas.openxmlformats.org/wordprocessingml/2006/main'
}

with open(output_file, 'w', encoding='utf-8') as out:
    for name in templates:
        path = os.path.join(public_dir, name)
        out.write(f"\n=========================================\nFILE: {name}\n=========================================\n")
        if not os.path.exists(path):
            out.write("File does not exist!\n")
            continue
        
        # Check if zipfile (docx format)
        if not zipfile.is_zipfile(path):
            out.write("Not a zipfile (likely binary .doc format).\n")
            # Try to read simple strings if it's a binary file
            try:
                with open(path, 'rb') as f:
                    content = f.read()
                # Very crude binary string extraction
                text_runs = []
                current = []
                for b in content:
                    if 32 <= b < 127 or b > 127: # printable or non-ascii
                        current.append(bytes([b]))
                    else:
                        if len(current) > 5:
                            try:
                                s = b''.join(current).decode('utf-8', errors='ignore')
                                if s.strip():
                                    text_runs.append(s.strip())
                            except:
                                pass
                        current = []
                out.write(f"Extracted crude text ({len(text_runs)} runs):\n")
                out.write(" ".join(text_runs[:50]) + "\n")
            except Exception as e:
                out.write(f"Error reading binary file: {e}\n")
            continue

        try:
            with zipfile.ZipFile(path) as z:
                doc_xml = z.read("word/document.xml")
                root = ET.fromstring(doc_xml)
                
                for elem in root.iter():
                    if elem.tag.endswith('p'):
                        text = ""
                        for t in elem.findall('.//w:t', namespaces):
                            if t.text:
                                text += t.text
                        if text.strip():
                            align = elem.find('.//w:jc', namespaces)
                            align_val = align.attrib.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}val', '') if align is not None else 'left'
                            out.write(f"[{align_val}] {text}\n")
        except Exception as e:
            out.write(f"Error reading docx XML: {e}\n")

print("Templates dumped successfully to templates_dump.txt")

File: scratch/inspect_all_templates.js
import fs from 'fs';
import JSZip from 'jszip';
import path from 'path';

const publicDir = 'd:/CBSV/public';
const files = [
  '1. Mau 10_KND_Ban tu kiem diem_DHKT_2025.docx',
  '5. Biên bản họp Chi Đoàn.docx',
  '6. Biên bản họp lớp.docx'
];

async function inspect() {
  for (const filename of files) {
    const filePath = path.join(publicDir, filename);
    if (!fs.existsSync(filePath)) continue;
    
    const buffer = fs.readFileSync(filePath);
    const zip = await JSZip.loadAsync(buffer);
    const docXml = await zip.file('word/document.xml').async('string');
    
    console.log(`\n==================================================`);
    console.log(`FILE: ${filename}`);
    
    for (const key of ['uu_diem', 'khuyet_diem']) {
      const idx = docXml.indexOf(`{{${key}}}`);
      if (idx !== -1) {
        // Find the containing <w:p> ... </w:p>
        const pStart = docXml.lastIndexOf('<w:p ', idx);
        const pEnd = docXml.indexOf('</w:p>', idx) + 6;
        if (pStart !== -1 && pEnd !== -1) {
          console.log(`--- Context for {{${key}}} ---`);
          console.log(docXml.substring(pStart, pEnd));
        }
      }
    }
  }
}

inspect().catch(console.error);
